presto-raptorx/src/main/java/com/facebook/presto/raptorx/systemtable/TableSystemTable.java

```diff
@@ -0,0 +1,174 @@
+    {
+        this.transactionManager = requireNonNull(transactionManager, "transactionManager is null");
+
+        Type varcharArray = typeManager.getType(parseTypeSignature("array<varchar>"));
+        this.tableMetadata = new ConnectorTableMetadata(
+                new SchemaTableName("system", "tables"),
+                ImmutableList.<ColumnMetadata>builder()
+                        .add(new ColumnMetadata("table_schema", VARCHAR))
+                        .add(new ColumnMetadata("table_name", VARCHAR))
+                        .add(new ColumnMetadata("temporal_column", VARCHAR))
+                        .add(new ColumnMetadata("ordering_columns", varcharArray))
+                        .add(new ColumnMetadata("distribution_name", VARCHAR))
+                        .add(new ColumnMetadata("bucket_count", BIGINT))
+                        .add(new ColumnMetadata("bucketing_columns", varcharArray))
+                        .build());
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return tableMetadata;
+    }
+
+    @Override
+    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        Transaction transaction = transactionManager.get(transactionHandle);
+
+        Map<Integer, NullableValue> values = extractFixedValues(constraint).orElse(ImmutableMap.of());
+        Optional<String> schemaName = getStringValue(values.get(getColumnIndex(tableMetadata, "table_schema")));
+        Optional<String> tableName = getStringValue(values.get(getColumnIndex(tableMetadata, "table_name")));
+
+        // TODO: fetch table info in bulk
+        List<TableInfo> tables = listTables(transaction, schemaName, tableName).stream()
+                .map(transaction::getTableId)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .map(transaction::getTableInfo)
+                .collect(toImmutableList());
+
+        PageListBuilder builder = PageListBuilder.forTable(tableMetadata);
+
+        for (TableInfo table : tables) {
+            builder.beginRow();
+            writeTable(builder, transaction, table);
+            builder.endRow();
+        }
+
+        return new FixedPageSource(builder.build());
+    }
+
+    private static void writeTable(PageListBuilder builder, Transaction transaction, TableInfo table)
+    {
+        // extract special columns
+        Optional<String> temporalColumn = table.getTemporalColumn()
+                .map(ColumnInfo::getColumnName);
+
+        List<String> ordering = table.getSortColumns().stream()
+                .map(ColumnInfo::getColumnName)
+                .collect(toImmutableList());
+
+        List<String> bucketing = table.getBucketColumns().stream()
```

Comment:
Move to where it is used (`else` branch)?

Reply:
yeah